File: desktop/desktop-sdk/src/api/practice.ts
/**
 * Practice-mode (typing-drill) API surface.
 *
 * Backed by the in-process `PracticeService` on the Tauri side — every
 * type below is re-exported from the specta-generated bindings so the
 * renderer's `@shared/practice` shape and the Rust DTO shape stay
 * locked in step.
 *
 * Channel names follow the snake_case convention established by #109.
 * Struct args use the `{ args }` envelope; scalar args (a single
 * `spaceId`) ride flat — matching the convention in
 * `documents.ts` / `spaces.ts`.
 */

import type {
	Exercise,
	ExerciseAttempt,
	ExerciseDraft,
	GenerateExerciseInput,
	RecordSessionResponse,
} from "../bindings";
import type { Transport } from "../transport";

export type {
	Exercise,
	ExerciseAttempt,
	ExerciseDifficulty,
	ExerciseDraft,
	ExerciseDraftMetadata,
	ExerciseMetadata,
	ExerciseSource,
	GenerateExerciseInput,
	LeaderboardEntry,
	RecordSessionResponse,
} from "../bindings";

export function practice(t: Transport) {
	return {
		listExercises: (spaceId: string) => t.invoke<Exercise[]>("practice_list_exercises", { spaceId }),
		saveExercise: (draft: ExerciseDraft) => t.invoke<Exercise>("practice_save_exercise", { args: draft }),
		recordSession: (attempt: ExerciseAttempt) =>
			t.invoke<RecordSessionResponse>("practice_record_session", { args: attempt }),
		generateExercise: (input: GenerateExerciseInput) =>
			t.invoke<ExerciseDraft>("practice_generate_exercise", { args: input }),
	};
}

File: desktop/desktop-commands/src/practice.rs
//! Tauri presenter for `desktop_api::practice::*`.

use desktop_api::{
    AppState,
    practice::{self as api, Exercise, ExerciseAttempt, ExerciseDraft, GenerateExerciseInput, RecordSessionResponse},
};
use desktop_core::error::DesktopResult;
use tauri::State;

#[tauri::command]
#[specta::specta]
pub async fn practice_list_exercises(
    state: State<'_, AppState>,
    space_id: Option<String>,
) -> DesktopResult<Vec<Exercise>> {
    api::list_exercises(state.inner(), space_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn practice_save_exercise(state: State<'_, AppState>, args: ExerciseDraft) -> DesktopResult<Exercise> {
    api::save_exercise(state.inner(), args).await
}

#[tauri::command]
#[specta::specta]
pub async fn practice_record_session(
    state: State<'_, AppState>,
    args: ExerciseAttempt,
) -> DesktopResult<RecordSessionResponse> {
    api::record_session(state.inner(), args).await
}

#[tauri::command]
#[specta::specta]
pub async fn practice_generate_exercise(
    state: State<'_, AppState>,
    args: GenerateExerciseInput,
) -> DesktopResult<ExerciseDraft> {
    api::generate_exercise(state.inner(), args).await
}
